### PR TITLE
Add "auto" option to vd-lavc-dr and make it the default

### LINKED ISSUE #10972: massive cpu increase with v0.35 (cause: DR on Intel iGPU)
### Important Information

Provide following Information:
- mpv version: 0.35.0-2
- Linux Distribution and Version: Arch linux
- Source of the mpv binary: Arch community repo
- If known which version of mpv introduced the problem: 0.35
- Window Manager and version: xfwm4 4.16.1
- GPU driver and version:  xf86-video-intel 2.99.917+916+g31486f40-2
- Possible screenshot or video of visual glitches: N/A
- CPU: Intel core i3-350M (2C/4T) 2.266 GHz max

### Reproduction steps
no-config option has no effect on the problem

### Expected behavior

Before version 0.35 mpv played 1080p30 x264 files with no problem, cpu usage was between 40-60%
This is without hw acceleration. With hw acceleration, problem disappears.

### Actual behavior

After v0.35 1080p30 x264 can no longer play at required fps, cpu usage immediately jumps to 100%

Reverting to v0.34.1-5 fixes the problem.

### Log file

[mpv-log.txt](https://github.com/mpv-player/mpv/files/10169608/mpv-log.txt)

### Sample files

any 1080p (or near) x264 video will do

### LINKED ISSUE #10997: Poor performance when using ARB_buffer_storage on ARM platforms with GL output
### Important Information

Provide following Information:
- mpv version: 0.35.0
- Linux Distribution and Version: Archlinux ARM
- Source of the mpv binary: Archlinux ARM packages
- If known which version of mpv introduced the problem: N/A
- Window Manager and version: weston-11.0.0
- GPU driver and version: Mesa-22.2
- Possible screenshot or video of visual glitches: N/A

### Reproduction steps

Try playing any video with mpv on a platform with Lima driver (however at least vc4 and panfrost are likely affected as well). Performance is terrible, mpv consumes a lot of CPU cycles.
Disabling ARB_buffer_storage with `MESA_EXTENSION_OVERRIDE=-GL_ARB_buffer_storage` results in significantly lower CPU usage.

I profiled it with perf, and it shows that most of the time is spent in libavcodec.

Apparently mpv uses directly mapped persistent coherent buffers with a decoder and assumes that the buffer is cached, but it isn't always true - it will return write-combined buffer on most ARM platforms (e.g. with vc4, panfrost, lima drivers)

It isn't a driver bug, since GL_ARB_buffer_storage spec doesn't specify whether directly mapped buffer should be cached or uncached.

Please consider adding a command line (and/or config) option to disable GL_ARB_buffer_storage usage. You may want to enable it by default on ARM platforms.

As an alternative, you can try using a staging buffer allocated with malloc for software decoder and memcpy the data into GPU buffer, it will be more efficient than using TexSubImage2D for uploading the data. Keep in mind that VPU-accelerated decoding doesn't need a staging buffer, it's only needed for software decoding.

See https://gitlab.freedesktop.org/mesa/mesa/-/issues/7862 for additional details

### Expected behavior

CPU usage is on par with TexSubImage2D code path.

### Actual behavior

CPU usage is 4x higher

### Log file

Please see https://gitlab.freedesktop.org/mesa/mesa/-/issues/7862 for log files and additional details

### PR DESCRIPTION
**Pull request by CounterPillow**
This reverts commit d6af6efbf9873906bd21000551fa0f284c441eff.

On platforms where the GPU cannot cache snoop the buffers we're rendering into, DR makes the decoding horrifically slow. Some of the affected platforms are any ARM with either lima, panfrost or vc4, (A64, RK3xxx, RPis, etc.) as well as seemingly low-end Intel SoCs according to user reports.

Defaulting this option to on is therefore a potential show stopper for many people who are unaware of it, as it makes the playback unusable. Meanwhile, on systems where DR works (like mine), the performance benefit appears to be in the realm of 10%, which is fine to miss out on if a user doesn't know about the option.

What seemingly changed to make this problem crop up as of late is that both mpv and drivers have gotten better, and DR is used more often.

Closes #10997
Closes #10972

### CODE REVIEW

**Comment by haasn** [comment]
Propose also to add `--vd-lavc-dr=auto` which auto-enables it on exactly vulkan.

**Comment by sfan5** [comment]
`<sfan5> back to the original point: can we measure on which gpu/platform combos DR actually has an advantage?`

If it turns out that it's still ok on OpenGL/NVIDIA and OpenGL/AMD then we could add an `auto` option that enables it only there.

--

By the way, what was the reasoning why Vulkan is not affected by the original issue?

**Comment by aufkrawall** [comment]
> If it turns out that it's still ok on OpenGL/NVIDIA and OpenGL/AMD then we could add an `auto` option that enables it only there.

Can we add iris GL driver (>=Haswell IGP) to that list?

**Comment by CounterPillow** [comment]
> Can we add iris GL driver (>=Haswell IGP) to that list?

No, because this option wreaks havoc on some lower end Intel iGPUs that AFAIK use that same driver.

**Comment by aufkrawall** [comment]
> No, because this option wreaks havoc on some lower end Intel iGPUs that AFAIK use that same driver.

Are we sure these aren't stuck on legacy i965 driver? Apparently, `vd-lavc-dr=yes` never caused any issue on my lowest-end Gemini Lake Gen 9.5 IGP.

**Comment by CounterPillow** [comment]
As far as I know, the problem with intel is not related to the age of the specific platform, but the specific implementation. IIRC we've seen modern Centrino/Pentium stuff run into problems. E.g. the Intel Pentium N4200 which afaik uses the modern driver has this issue.

**Comment by sfan5** [comment]
Here's the blacklist approach I once came up with once that detects iGPUs that _should_ have this issue: https://github.com/sfan5/mpv/commit/9502b4645ad219a350318bd6a75410b530c8655b#diff-4ae7c5e39ff99c9b99e74e2e5597f2751f446d753f27f34c1c6bc26977c1b8b3
Too fragile to be actually used though, better would be to whitelist by vendor (so AMD and NVIDIA) if the facts match.

**Comment by sfan5** [comment]
Okay I collected some data from the machines I had available here and measured FPS according to this https://haasn.dev/posts/2017-10-05-how-to-benchmark-mpvs-raw-throughput.html.

Tested on Xfce with xfwm4 enabled as compositor, I did not resize the mpv window, video was 720p H.264 so shouldn't bottleneck.
<pre>
Radeon 6500 XT on X11 via 'amdgpu' driver (Mesa 22.3.1)
~1800 without, ~2400 with

HD Graphics 4600 (HSW GT2) on X11 via 'modesetting' driver (Mesa 22.3.1)
~380 without, ~370 with

HD Graphics 4600 (HSW GT2) on X11 via 'intel' driver
~390 without, ~380 with (margin of error?)

GeForce GTX 1050 Ti on X11 via 'nvidia' driver (NVIDIA 525.60.11)
~1100 without, ~1500 with

HD Graphics 405 (BSW) on X11 via 'intel' driver (Mesa 22.2.4) (*)
~290 without, ~90 with

Xe Graphics (TGL GT2) on X11 via 'modesetting' driver (Mesa 22.3.1) (*)
~2000 without, ~850 with

AMD ARUBA (A8-4500M) on X11 via 'modesetting' driver (Mesa 22.3.1) (*)
~430 without, ~470 with

(*) = Laptop
</pre>

In short it looks like we can safely enable DR for AMD and NVIDIA.

**Comment by CounterPillow** [comment]
I've force-pushed the branch with a new approach based on sfan5's past work which adds an auto flag and only enables it on AMD, nvidia and nouveau (lol). kept him as the commit author since most of the commit is his, but added myself as co-authored-by to make it clear that any breakage is probably my fault.

File: DOCS/man/options.rst [hunk 1]
```diff
@@ -1701,15 +1701,18 @@ Video
...
+    Enable direct rendering (default: auto). If this is set to ``yes``, the
     video will be decoded directly to GPU video memory (or staging buffers).
     This can speed up video upload, and may help with large resolutions or
     slow hardware. This works only with the following VOs:
 
         - ``gpu``: requires at least OpenGL 4.4 or Vulkan.
         - ``libmpv``: The libmpv render API has optional support.
 
+    The ``auto`` option will try to guess whether DR is fast or slow on your
+    particular hardware.
```
Comment (sfan5) [inline DOCS/man/options.rst:1714]:
```suggestion
    The ``auto`` option will try to guess whether DR can improve performance
    on your particular hardware. Currently this enables it on AMD or NVIDIA
    if using OpenGL or unconditionally if using Vulkan.
```

File: video/out/opengl/common.c [hunk 2]
```diff
@@ -650,6 +663,11 @@ void mpgl_load_functions2(GL *gl, void *(*get_fn)(void *ctx, const char *n),
         mp_verbose(log, "Detected suspected software renderer.\n");
     }
 
+    if (!is_fast_dr(gl)) {
+        gl->mpgl_caps |= MPGL_CAP_SLOW_DR;
+        mp_verbose(log, "Did not detect GPU with fast DR, disabling.\n");
```
Comment (sfan5) [inline video/out/opengl/common.c:668]: I'd move this to vd_lavc.c so it's only printed when actually relevant

File: video/out/vo.h [hunk 3]
```diff
@@ -128,6 +128,10 @@ enum mp_voctrl {
     VOCTRL_GET_DISPLAY_RES,             // int[2]
     VOCTRL_GET_WINDOW_ID,               // int64_t*
 
+    // Return false if VO is reasonably sure that DR would be bad for
+    // performance even if it works.
+    VOCTRL_GET_WANTS_DR,                // bool*
```
Comment (sfan5) [inline video/out/vo.h:133]: I'd rename this to `VOCTRL_GET_DR_HINT`

File: video/out/vo_gpu.c [hunk 4]
```diff
@@ -224,6 +224,9 @@ static int control(struct vo *vo, uint32_t request, void *data)
         p->ctx->fns->reconfig(p->ctx);
         resize(vo);
         return true;
+    case VOCTRL_GET_WANTS_DR:
+        *(bool*) data = !(p->ctx->ra->caps & RA_CAP_SLOW_DR);
+        return true;
```
Comment (sfan5) [inline video/out/vo_gpu.c:229]: `gpu_next` will need this too, but not sure what it should look like.

the only other vo with this is `libmpv` which I guess should just duplicate this.

**Review by sfan5 — CHANGES_REQUESTED** [review]

File: video/out/vo_gpu.c (hunk 4 above)
Comment (haasn) [reply video/out/vo_gpu.c:229]: https://code.videolan.org/videolan/libplacebo/-/merge_requests/336

File: video/decode/vd_lavc.c [hunk 5]
```diff
@@ -696,6 +697,13 @@ static void init_avctx(struct mp_filter *vd)
         mp_set_avcodec_threads(vd->log, avctx, lavc_param->threads);
     }
 
+    if (lavc_param->dr == -1) {
+        bool wants_dr = false;
```
Comment (sfan5) [inline video/decode/vd_lavc.c:701]: another (minor) thing: if the vo doesn't have an answer for us we may want to default to true

File: video/decode/vd_lavc.c [hunk 6]
```diff
@@ -696,6 +697,15 @@ static void init_avctx(struct mp_filter *vd)
         mp_set_avcodec_threads(vd->log, avctx, lavc_param->threads);
     }
 
+    if (lavc_param->dr == -1) {
+        bool wants_dr = true;
+        if (ctx->vo)
+            vo_control(ctx->vo, VOCTRL_GET_DR_HINT, &wants_dr);
+        if (!wants_dr)
+            mp_verbose(vd->log, "Did not detect GPU with fast DR, disabling.\n");
+        lavc_param->dr = !!wants_dr ? 1 : 0;
```
Comment (sfan5) [inline video/decode/vd_lavc.c:706]: can be either `!!wants_dr ? 1 : 0` or `wants_dr ? 1 : 0`

File: video/out/vo_gpu.c (hunk 4 above)
Comment (sfan5) [reply video/out/vo_gpu.c:229]: @CounterPillow still missing for libmpv

**Comment by haasn** [comment]
@CounterPillow Rather than bothering with a VOCTRL just to query a single capability, I feel like we should just make `get_image` fail in this case. The calling code will already transparently disable DR if get_image fails. We can also extend its signature by an `int flags` and add a flag like `DR_NO_SLOW` or `DR_FORCE` or w/e, to allow a distinction between `auto` and `yes` (or maybe it should be `yes` and `force`?)

**Comment by haasn** [comment]
I tried reimplementing this based on my idea, but it turned out very ugly too, because of many places that need API changes.

I would also prefer splitting up the commit before merging, but does it look good as-is?

**Review by sfan5 — APPROVED** [review]
